### PR TITLE
HIVE-27324: Hive query with NOT IN condition is giving incorrect results when the sub query table contains the null value

### PR DESCRIPTION
**Pull request by Diksha628**
JIRA Link: [https://issues.apache.org/jira/browse/HIVE-27324](url)

### What changes were proposed in this pull request?

This PR proposes changes that when there is a NOT IN operator in the query, and cbo is disabled, for e.g.:
select * from outer_query where (outer_query.outerQuery_col NOT IN (select distinct(subq_col)  subquery_table); there be a check for outer_query.outerQuery_col is not NULL. 

### Why are the changes needed?

As mentioned in the JIRA, a null NOT IN <x,y,.....> case will be encountered and it should return false, that we can see in cbo enabled behaviour. If the changes proposed are not there, the output will contain NULL values in the col referenced, as well.

### Does this PR introduce _any_ user-facing change?

Yes, 
Earlier, as can be seen in the JIRA, when cbo is disabled,  a ( null not in {x,y,.....}  returns true , but after the changes it returns false, as it should as per sql standards and as done in cbo enabled case.

### Is the change a dependency upgrade?

No

### How was this patch tested?

A test file has been added to MiniLlapCliDriver to include tests containing null not in cases. The tests have been tested against output given for the same when cbo is enabled.

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 1]
```diff
@@ -3742,6 +3755,80 @@ private Operator genNotNullFilterForJoinSourcePlan(QB qb, Operator input,
     return output;
   }
 
+  /** This function is used for the purpose of adding an
+   *  outerQueryTable.outerQueryCol is not null predicate to the join,
+   *  since it is not added naturally like it is done in the case of "IN" operator
+   **/
+  private Operator genNotNullFilterForJoinSourcePlanNIC(QB qb, Operator input,
```
Comment (sankarh) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:3762]: Too many duplicate code. Can simplify this.

File: ql/src/test/queries/clientpositive/notInTest.q [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+INSERT OVERWRITE TABLE t5
+  select *
+  where b.age in
+   (select c.age
+    from t3 c
+   )
+;
+
+drop table if exists t3;
```
Comment (sankarh) [inline ql/src/test/queries/clientpositive/notInTest.q:28]: Can run the queries on these tables itself instead of dropping and re-creating.

File: ql/src/test/queries/clientpositive/notInTest.q [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+create table t4 (id int,name string, age int);
+insert into t4 values(1,'Sagar',23),(3,'Surya',23),(4,'Raman',45),(5,'Scott',23),(6,'Ramya',5),(7,'',23),(8,'',23);
+
+create table t5 (id int,name string, ages int);
+insert into t5 values(1,'Sagar',23),(3,'Surya',NULL),(4,'Raman',45),(5,'Scott',23),(6,'Ramya',5),(7,'',23),(8,'',23);
+
+set hive.cbo.enable = false;
+
+select count(*) from t3
```
Comment (sankarh) [inline ql/src/test/queries/clientpositive/notInTest.q:43]: Output the values instead of just count for all queries.

File: ql/src/test/queries/clientpositive/notInTest.q [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+where age not in (select distinct(age) age from t4);
+
+select count(*) from t3
```
Comment (sankarh) [inline ql/src/test/queries/clientpositive/notInTest.q:46]: Try this non in with String data type.

**Review by sankarh — CHANGES_REQUESTED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 5]
```diff
@@ -3603,7 +3603,20 @@ private Operator genFilterPlan(ASTNode searchCond, QB qb, Operator input,
          * push filters only for this QBJoinTree. Child QBJoinTrees have already been handled.
          */
         pushJoinFilters(qb, joinTree, aliasToOpInfo, false);
-        input = genJoinOperator(qbSQ, joinTree, aliasToOpInfo, input);
+        boolean notInCheck = (subQuery.getNotInCheck() != null)?true : false;
+        if(notInCheck && !qb.isMultiDestQuery()){
```
Comment (sankarh) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:3607]: There is already a code flow at line 3590 that handles notInCheck. See if this code flow can be combined there.

Comment (Diksha628) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:3607]: The code block with the notInCheck at that line, handles the not in operator such that if subquery table.col_referenced has a null value, output will be empty. The code block added in this PR aligns with the behaviour of queries with "IN"  operator, where the null not in case is handled during join time by adding a predicate.

File: ql/src/test/queries/clientpositive/notInTest.q (hunk 3 above)
Comment (Diksha628) [reply ql/src/test/queries/clientpositive/notInTest.q]: changed this

File: ql/src/test/queries/clientpositive/notInTest.q (hunk 4 above)
Comment (Diksha628) [reply ql/src/test/queries/clientpositive/notInTest.q]: added in the tests

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 1 above)
Comment (Diksha628) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: done

**Comment by Diksha628** [comment]
Hi @zabetak , can you please review this PR?

**Comment by zabetak** [comment]
@Diksha628 Sorry, I am pretty busy this period probably will not make it in time.

File: ql/src/test/queries/clientpositive/notInTest.q [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+
+select * from t3
+where age not in (select distinct(ages) ages from t5 )
+order by age ;
+
+select count(*) from t3
+where age not in (23,22, null );
+
+explain select * from t3
```
Comment (sankarh) [inline ql/src/test/queries/clientpositive/notInTest.q:33]: Run explain query on other query with sub-query on t5 as well.

File: ql/src/test/queries/clientpositive/notInTest.q [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
```
Comment (sankarh) [inline ql/src/test/queries/clientpositive/notInTest.q:61]: Run explain on select query on sub-query with t5.

Comment (Diksha628) [reply ql/src/test/queries/clientpositive/notInTest.q:67]: done, added the explain plan.

File: ql/src/test/queries/clientpositive/notInTest.q (hunk 6 above)
Comment (Diksha628) [reply ql/src/test/queries/clientpositive/notInTest.q:33]: done, added explain plan for the other subquery as well.

**Comment by douglasawh** [comment]
Hi folks, is there any sort of ETA for when this might be fixed? Q1 next year, perhaps?

File: ql/src/test/results/clientpositive/llap/special_character_in_tabnames_quotes_1.q.out [hunk 8]
```diff
@@ -3071,23 +3086,49 @@ STAGE PLANS:
...
+                    Select Operator
+                      expressions: p_mfgr (type: string), p_size (type: int)
+                      outputColumnNames: _col0, _col1
+                      Statistics: Num rows: 5 Data size: 510 Basic stats: COMPLETE Column stats: COMPLETE
+                      Group By Operator
+                        aggregations: min(_col1)
+                        keys: _col0 (type: string)
+                        minReductionHashAggr: 0.4
+                        mode: hash
+                        outputColumnNames: _col0, _col1
```
Comment (maheshk114) [inline ql/src/test/results/clientpositive/llap/special_character_in_tabnames_quotes_1.q.out:3123]: can you please check why the extra group by operator is added ?

**Review by maheshk114 — CHANGES_REQUESTED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 9]
```diff
@@ -3603,7 +3603,29 @@ private Operator genFilterPlan(ASTNode searchCond, QB qb, Operator input,
...
+         * )
+         * order by age
+         * INSERT OVERWRITE TABLE t5
+         * select *
+         * where b.age in
+         * (select c.age
+         * from t3 c
+         * );
+         */
+        boolean notInCheck = (subQuery.getNotInCheck() != null && !qb.isMultiDestQuery() );
```
[the remark this answers is not in the document]
Comment (maheshk114) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:3727]: nit : changed the name to notInCheckPresent

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 10]
```diff
@@ -9909,7 +9952,13 @@ private Operator genJoinOperator(QB qb, QBJoinTree joinTree,
       // generate a ReduceSink operator for the join
       String[] srcs = baseSrc[i] != null ? new String[] {baseSrc[i]} : joinTree.getLeftAliases();
       if (!isCBOExecuted()) {
-        srcOps[i] = genNotNullFilterForJoinSourcePlan(qb, srcOps[i], joinTree, joinKeys[i]);
+        /*
+         * The condition srcOps[i] == OuterSrcOp is used to make sure that the predicate for notnull check
+         * is added only for the outer query table.outerqueryCol
+         * even after the outer join condition
+         */
+        boolean outerNotInCheck = ( notInCheck && (srcOps[i] == OuterSrcOp) );
```
Comment (maheshk114) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java:9960]: nit : remove extra space after OuterSrcOp)

File: ql/src/test/results/clientpositive/llap/special_character_in_tabnames_quotes_1.q.out (hunk 8 above)
Comment (Diksha628) [reply ql/src/test/results/clientpositive/llap/special_character_in_tabnames_quotes_1.q.out:3123]: There are now extra group by operators, because after adding one more filter, the reducer 5 is now divided into 2 reducers. Both sharing the same parent operator, which makes it look like an extra group by is added.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java [hunk 11]
```diff
@@ -3603,7 +3603,29 @@ private Operator genFilterPlan(ASTNode searchCond, QB qb, Operator input,
          * push filters only for this QBJoinTree. Child QBJoinTrees have already been handled.
          */
         pushJoinFilters(qb, joinTree, aliasToOpInfo, false);
-        input = genJoinOperator(qbSQ, joinTree, aliasToOpInfo, input);
+
+        /*
+         *  Note that: in case of multi dest queries, with even one containing a notIn operator, the code is not changed yet.
+         *  That needs to be worked on as a separate bug
```
[the remark this answers is not in the document]
Comment (Diksha628) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: done

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Diksha628) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: done

File: ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java (hunk 10 above)
Comment (Diksha628) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/SemanticAnalyzer.java]: done

**Review by maheshk114 — APPROVED** [review]

**Comment by maheshk114** [comment]
+1 Looks good to me

**Review by sankarh — APPROVED** [review]
